Request expert feature

paulocheque opened the pull request:
This PR is not complete yet. I am still working on it.

paulocheque committed:
Updated Flask-OAuthlib lib to 0.9.2 version

paulocheque committed:
Option to the writer connect his/her hack.hands() account

durden commented:
Looks like a good start, just a few comments:
1. What does upgrading the oauthlib get us?
   - It's no problem to upgrade. I'm just curious what the benefits are so we can have it documented here.
2. I think we should keep the hack hands stuff separated from the main views and remote.py.
   - The reasoning here is the hack hands stuff will most likely be used by very few users. We can make it slightly easier to navigate the code by separating this off too.  This way potential contributors don't have to worry about this integration at all unless they specifically need it.
   - Maybe we should make a remote sub package and put in a `github.py` and a `hackhands.py` file?
   - The easier option is to put all the hack hands views in `hackhands_views.py` and the remote oauth stuff in `hackhands.py`.
     - These are just examples so let me know if you have preferences on the naming or layout.

paulocheque committed:
Moving hack.hands code to a separated module. Completing the connection view function.

paulocheque committed:
Saving in the cache the hack.hands user information. Fixed the menu link to hack.hands profile page. Fixed session information/login/logout.

paulocheque commented:
Hi,
1. I was stucked with some strange erros during the OAuth requests to the hack.hands API, so I decide to check the Flask-oauthlib for issues.
   In fact, the errors I was facing was not related to the lib, but since this new version had some fixes (https://github.com/lepture/flask-oauthlib/blob/master/CHANGES.rst), I updated to advance my research.
2. I moved to a separated module (`hackhands.py`). But there are some lines that need to be in the core of the system: like some session variables.

paulocheque committed:
Saving hack.hands data forever, without expiring time.

paulocheque commented:
In the logout view, why dont we just call `session.clear()`?

durden commented:
Good point. There's no reason to not call clear. Only downside is clearing
the previously viewed page if we need a redirect but even that isn't very
useful.

On Wednesday, May 18, 2016, Paulo Cheque notifications@github.com wrote:

> In the logout view, why dont we just call session.clear()?
> 
> —
> You are receiving this because you commented.
> Reply to this email directly or view it on GitHub
> https://github.com/pluralsight/guides-cms/pull/77#issuecomment-220149062

paulocheque committed:
Consider user_info.html as a closed component.

durden commented:
Sounds good. Upgrading isn't a problem I was just curious if there was a
critical doc we needed to make note of. Thanks for separating out the
hackhands stuff. It shouldn't be a problem if the hackhands integration
puts something in the main app config or the main app context in
**init**.py, etc.

On Wednesday, May 18, 2016, Paulo Cheque notifications@github.com wrote:

> In the logout view, why dont we just call session.clear()?
> 
> —
> You are receiving this because you commented.
> Reply to this email directly or view it on GitHub
> https://github.com/pluralsight/guides-cms/pull/77#issuecomment-220149062

paulocheque committed:
Button "Hire me on hack.hands()" in the article and profile pages.

paulocheque committed:
A cleaner and safer logout function

paulocheque commented:
I guess the feature has been done. You can check on the dev env.

# Important Considerations

## New env vars:

To deploy in a new environment (including the production), we need to define the new env vars on Heroku:

```
heroku config:set HACK_HANDS_BASE_URL="..." --app ...
heroku config:set HACK_HANDS_CLIENT_ID="..." --app ...
heroku config:set HACK_HANDS_CLIENT_SECRET="..." --app ...
```

## Authorization / Config

We only configured the staging env for a while. Before deploying for production, we need to configure hack.hands() for the new env.

## Storage

I am using Redis to store this information. I setted the `timeout` to `None` so this information will never expire. But we need to consider the Redis as a true storage from now on. I mean, we cannot clear the `cache` of this new information.

paulocheque committed:
Better redis key separator

durden commented on pskb_website/views.py:
```diff
@@ -164,6 +165,29 @@ def authorized():
+@app.route('/hackhands_login')
+def hackhands_login():
+    """hack.hands() oauth2 authorization"""
+    return remote.hackhands.authorize(callback='http://guides-dev.herokuapp.com/auth/hackhands', _external=True)
```
We should use `url_for` here so we don't hard-code the domain, etc.

durden commented:
I looked through this and looks good.  I've added 1 comment to the diff:

[Using url_for](https://github.com/pluralsight/guides-cms/pull/77/files/46c767341d1020ba5d171ee502876da1a2466ebc..d3e9d2404d8d033c2179685e665b2bbb9dda47c1#r64361102)

durden commented:
I think we should hold off on this feature until the new design is pushed to production and fully tested.  I don't think this feature will change once the new design is in place, but it's easier to integrate one big change at a time.  Also, we can rebase this feature after the new design changes are in and re-test everything.

paulocheque committed:
Removed hard coded urls

paulocheque commented:
Nice catch! Fixed.

durden commented:
Is your [test site](http://guides-dev.herokuapp.com) running this version?  If so, how do I attempt to hook up the hack hands feature?

durden commented:
We need to consider the workflow for this once we roll it into production.  Maybe we could put a link at the top of the profile page to link a hackhands account?  This way each time a user logs in they see the link, but it's not intrusive.  Of course, the link should only show up if the user is viewing their own profile page.

paulocheque commented:
I deployed into the [test site](http://guides-dev.herokuapp.com/) this feature. Right now, I added a link into the navbar. I can move the link to another place after feedbacks.

durden commented:
@paulocheque Thanks for deploying it for testing!  Just for historical reference here's what the feature looks like now:

<img width="1264" alt="screenshot 2016-06-14 10 46 59" src="https://cloud.githubusercontent.com/assets/58063/16036732/88669d28-321d-11e6-9c8b-987a07da1d6b.png">

I vote we move the 'connect with hackhands link' to the dropdown menu under the users' avatar when they are logged in.  So that menu will then have 'my guides', 'connect to hackhands', and 'logout'.

We should also add a link to 'connect to hackhands' on the profile page in the following case:
- User is viewing their own profile
- User is logged in
- User doesn't already have the account linked

We could add this link at the top of the profile 'body' and use something like the CSS class `bg-info` to make the link show up as a 'banner' in a light blue color.  This is just a suggestion though so feel free to try some other styling if you have other ideas.

paulocheque committed:
Better places for the connection with hack.hands buttons: in the navbar dropdown menu and in the profile page for the current user.

paulocheque commented:
Cool! Added all changes. I deployed in the test site:

[1](http://guides-dev.herokuapp.com/author/durden)
[2](http://guides-dev.herokuapp.com/author/paulocheque)

The button colors are messy because I deployed both PRs in the same branch. But after the Design PR being accepted, the color should be fixed.

durden commented:
Looks great.  Let's finish up the design changes, rebase this branch, and then double-check the button colors.  Thanks!

paulocheque commented:
Rebased and fixed the button layout.

paulocheque commented:
Rebased to origin/master
